pkg/volume/flexvolume/probe.go

```diff
@@ -71,11 +71,20 @@ func (prober *flexVolumeProber) Init() error {
 // If probeAllNeeded is true, probe all pluginDir
 // else probe events in eventsMap
 func (prober *flexVolumeProber) Probe() (events []volume.ProbeEvent, err error) {
+	prober.mutex.RLock()
 	if prober.probeAllNeeded {
-		prober.testAndSetProbeAllNeeded(false)
-		return prober.probeAll()
+		prober.mutex.RUnlock()
+		prober.mutex.Lock()
+		// check again, if multiple readers got through the first if, only one should probeAll
+		if prober.probeAllNeeded {
+			events, err = prober.probeAll()
+			prober.probeAllNeeded = false
+			prober.mutex.Unlock()
+			return
+		}
+		prober.mutex.Unlock()
 	}
-
+	prober.mutex.RUnlock()
```

Comment:
While the code looks correct, it smells. Who will unlock the lock when `probeAll()` panics?

`probeAllNeeded` is set to true only in [Init()](https://github.com/kubernetes/kubernetes/blob/0746672679200985eae4d761834cfd8c39851f80/pkg/volume/flexvolume/probe.go#L58-L58). Can you remove `probeAllNeeded` and use `sync.Once` instead (created in Init()) and called here)?

Reply:
yeah that sounds much better, I've made the change to use `sync.Once`, but I need to keep `probeAllNeeded` as it's used [here](https://github.com/kubernetes/kubernetes/blob/0746672679200985eae4d761834cfd8c39851f80/pkg/volume/flexvolume/probe.go#L228)